File: src/dere_discord/daemon_agent.py
"""WebSocket client for the daemon's centralized agent API."""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator
from typing import Any

import websockets
from loguru import logger
from websockets.asyncio.client import ClientConnection

from dere_shared.agent_models import (
    ClientMessage,
    ClientMessageType,
    SessionConfig,
    StreamEvent,
    StreamEventType,
)


class DaemonAgentClient:
    """WebSocket client for daemon agent API.

    Provides a clean interface for Discord (and other clients) to interact
    with the centralized Claude agent service in the daemon.
    """

    def __init__(self, daemon_url: str = "ws://localhost:8787"):
        self._base_url = daemon_url.rstrip("/")
        self._ws: ClientConnection | None = None
        self._current_session_id: int | None = None
        self._lock = asyncio.Lock()

    @property
    def session_id(self) -> int | None:
        """Current session ID if connected."""
        return self._current_session_id

    @property
    def connected(self) -> bool:
        """Whether WebSocket is connected."""
        return self._ws is not None and self._ws.state.name == "OPEN"

    async def connect(self) -> None:
        """Establish WebSocket connection to daemon."""
        if self.connected:
            return

        async with self._lock:
            if self.connected:
                return

            ws_url = f"{self._base_url}/agent/ws"
            logger.debug("Connecting to daemon agent at {}", ws_url)

            self._ws = await websockets.connect(ws_url)
            logger.info("Connected to daemon agent WebSocket")

    async def close(self) -> None:
        """Close WebSocket connection."""
        async with self._lock:
            if self._ws:
                try:
                    await self._ws.close()
                except Exception as e:
                    logger.debug("Error closing WebSocket: {}", e)
                finally:
                    self._ws = None
                    self._current_session_id = None

    async def _send(self, msg: ClientMessage) -> None:
        """Send a message to the daemon."""
        if not self._ws:
            raise RuntimeError("Not connected to daemon")

        await self._ws.send(msg.to_dict().__str__().replace("'", '"'))

    async def _send_json(self, data: dict[str, Any]) -> None:
        """Send JSON data to the daemon."""
        if not self._ws:
            raise RuntimeError("Not connected to daemon")

        import json

        await self._ws.send(json.dumps(data))

    async def _receive(self) -> StreamEvent:
        """Receive a single event from the daemon."""
        if not self._ws:
            raise RuntimeError("Not connected to daemon")

        import json

        data = await self._ws.recv()
        if isinstance(data, bytes):
            data = data.decode("utf-8")

        parsed = json.loads(data)
        return StreamEvent.from_dict(parsed)

    async def new_session(self, config: SessionConfig) -> int:
        """Create a new agent session.

        Args:
            config: Session configuration

        Returns:
            Session ID

        Raises:
            RuntimeError: If session creation fails
        """
        await self.connect()

        msg = ClientMessage(type=ClientMessageType.NEW_SESSION, config=config)
        await self._send_json(msg.to_dict())

        event = await self._receive()
        if event.type == StreamEventType.SESSION_READY:
            self._current_session_id = event.data.get("session_id")
            logger.info("Created new agent session: {}", self._current_session_id)
            return self._current_session_id
        elif event.type == StreamEventType.ERROR:
            raise RuntimeError(f"Failed to create session: {event.data.get('message')}")
        else:
            raise RuntimeError(f"Unexpected response: {event.type}")

    async def resume_session(self, session_id: int) -> bool:
        """Resume an existing session.

        Args:
            session_id: ID of session to resume

        Returns:
            True if resumed successfully

        Raises:
            RuntimeError: If resume fails
        """
        await self.connect()

        msg = ClientMessage(type=ClientMessageType.RESUME_SESSION, session_id=session_id)
        await self._send_json(msg.to_dict())

        event = await self._receive()
        if event.type == StreamEventType.SESSION_READY:
            self._current_session_id = event.data.get("session_id")
            logger.info("Resumed agent session: {}", self._current_session_id)
            return True
        elif event.type == StreamEventType.ERROR:
            logger.warning("Failed to resume session {}: {}", session_id, event.data.get("message"))
            return False
        else:
            raise RuntimeError(f"Unexpected response: {event.type}")

    async def update_config(self, config: SessionConfig) -> bool:
        """Update current session configuration.

        Args:
            config: New session configuration

        Returns:
            True if updated successfully
        """
        if not self._current_session_id:
            raise RuntimeError("No active session to update")

        msg = ClientMessage(type=ClientMessageType.UPDATE_CONFIG, config=config)
        await self._send_json(msg.to_dict())

        event = await self._receive()
        if event.type == StreamEventType.SESSION_READY:
            logger.info("Updated session config")
            return True
        elif event.type == StreamEventType.ERROR:
            logger.warning("Failed to update config: {}", event.data.get("message"))
            return False
        else:
            raise RuntimeError(f"Unexpected response: {event.type}")

    async def query(self, prompt: str) -> AsyncIterator[StreamEvent]:
        """Send a query and stream response events.

        Args:
            prompt: User prompt to send

        Yields:
            StreamEvent objects as they arrive

        Raises:
            RuntimeError: If no active session
        """
        if not self._current_session_id:
            raise RuntimeError("No active session. Call new_session or resume_session first.")

        msg = ClientMessage(type=ClientMessageType.QUERY, prompt=prompt)
        await self._send_json(msg.to_dict())

        while True:
            event = await self._receive()
            yield event

            if event.type in (StreamEventType.DONE, StreamEventType.ERROR):
                if event.type == StreamEventType.ERROR and not event.data.get("recoverable", True):
                    break
                if event.type == StreamEventType.DONE:
                    break

    async def ensure_session(
        self,
        config: SessionConfig,
        session_id: int | None = None,
    ) -> int:
        """Ensure a session exists, creating or resuming as needed.

        Args:
            config: Session configuration
            session_id: Optional session ID to resume

        Returns:
            Active session ID
        """
        await self.connect()

        if session_id:
            if await self.resume_session(session_id):
                return self._current_session_id

        return await self.new_session(config)

File: packages/discord/src/daemon-agent.ts
const DEFAULT_WS_URL =
  process.env.DERE_DAEMON_WS_URL ??
  (process.env.DERE_DAEMON_URL
    ? process.env.DERE_DAEMON_URL.replace(/^http/, "ws")
    : "ws://localhost:8787");

export type StreamEventType =
  | "session_ready"
  | "text"
  | "tool_use"
  | "tool_result"
  | "thinking"
  | "error"
  | "done"
  | "cancelled"
  | "permission_request";

export type StreamEvent = {
  type: StreamEventType;
  data: Record<string, unknown>;
  timestamp?: number;
  seq?: number;
};

export type SessionConfig = {
  working_dir: string;
  output_style?: string;
  personality?: string | string[];
  model?: string | null;
  user_id?: string | null;
  allowed_tools?: string[] | null;
  include_context?: boolean;
  enable_streaming?: boolean;
  thinking_budget?: number | null;
  sandbox_mode?: boolean;
  sandbox_mount_type?: "direct" | "copy" | "none";
  sandbox_settings?: Record<string, unknown> | null;
  sandbox_network_mode?: "bridge" | "host";
  mission_id?: number | null;
  session_name?: string | null;
  auto_approve?: boolean;
  lean_mode?: boolean;
  swarm_agent_id?: number | null;
  plugins?: string[] | null;
  env?: Record<string, string> | null;
  output_format?: Record<string, unknown> | null;
};

type ClientMessageType =
  | "new_session"
  | "resume_session"
  | "query"
  | "update_config"
  | "ping"
  | "close";

type ClientMessage = {
  type: ClientMessageType;
  config?: SessionConfig;
  session_id?: number;
  prompt?: string;
  last_seq?: number;
};

function parseStreamEvent(raw: string): StreamEvent {
  const parsed = JSON.parse(raw) as StreamEvent;
  return {
    type: parsed.type,
    data: parsed.data ?? {},
    timestamp: parsed.timestamp ?? Date.now() / 1000,
    seq: parsed.seq ?? 0,
  };
}

async function waitForOpen(ws: WebSocket): Promise<void> {
  if (ws.readyState === WebSocket.OPEN) {
    return;
  }
  await new Promise<void>((resolve, reject) => {
    const onOpen = () => {
      cleanup();
      resolve();
    };
    const onError = () => {
      cleanup();
      reject(new Error("WebSocket connection failed"));
    };
    const onClose = () => {
      cleanup();
      reject(new Error("WebSocket closed before opening"));
    };
    const cleanup = () => {
      ws.removeEventListener("open", onOpen);
      ws.removeEventListener("error", onError);
      ws.removeEventListener("close", onClose);
    };
    ws.addEventListener("open", onOpen);
    ws.addEventListener("error", onError);
    ws.addEventListener("close", onClose);
  });
}

function waitForMessage(ws: WebSocket): Promise<StreamEvent> {
  return new Promise((resolve, reject) => {
    const onMessage = (event: MessageEvent) => {
      cleanup();
      try {
        const raw = typeof event.data === "string" ? event.data : String(event.data ?? "");
        resolve(parseStreamEvent(raw));
      } catch (error) {
        reject(error);
      }
    };
    const onError = () => {
      cleanup();
      reject(new Error("WebSocket error"));
    };
    const onClose = () => {
      cleanup();
      reject(new Error("WebSocket closed"));
    };
    const cleanup = () => {
      ws.removeEventListener("message", onMessage);
      ws.removeEventListener("error", onError);
      ws.removeEventListener("close", onClose);
    };

    ws.addEventListener("message", onMessage);
    ws.addEventListener("error", onError);
    ws.addEventListener("close", onClose);
  });
}

export class DaemonAgentClient {
  private readonly baseUrl: string;
  private ws: WebSocket | null = null;
  private currentSessionId: number | null = null;
  private connectLock: Promise<void> | null = null;

  constructor(baseUrl: string = DEFAULT_WS_URL) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  get sessionId(): number | null {
    return this.currentSessionId;
  }

  get connected(): boolean {
    return this.ws !== null && this.ws.readyState === WebSocket.OPEN;
  }

  async connect(): Promise<void> {
    if (this.connected) {
      return;
    }

    if (!this.connectLock) {
      this.connectLock = (async () => {
        if (this.connected) {
          return;
        }
        const wsUrl = `${this.baseUrl}/agent/ws`;
        const ws = new WebSocket(wsUrl);
        await waitForOpen(ws);
        this.ws = ws;
        console.info(`Connected to daemon agent WebSocket at ${wsUrl}`);
      })();
    }

    try {
      await this.connectLock;
    } finally {
      this.connectLock = null;
    }
  }

  async close(): Promise<void> {
    if (this.ws) {
      try {
        this.ws.close();
      } finally {
        this.ws = null;
        this.currentSessionId = null;
      }
    }
  }

  private async send(data: ClientMessage): Promise<void> {
    if (!this.ws) {
      throw new Error("Not connected to daemon");
    }
    this.ws.send(JSON.stringify(data));
  }

  private async receive(): Promise<StreamEvent> {
    if (!this.ws) {
      throw new Error("Not connected to daemon");
    }
    return waitForMessage(this.ws);
  }

  async newSession(config: SessionConfig): Promise<number> {
    await this.connect();
    await this.send({ type: "new_session", config });
    const event = await this.receive();
    if (event.type === "session_ready") {
      const sessionId = Number(event.data.session_id);
      if (!Number.isFinite(sessionId)) {
        throw new Error(`Invalid session id: ${String(event.data.session_id)}`);
      }
      this.currentSessionId = sessionId;
      return sessionId;
    }
    if (event.type === "error") {
      throw new Error(`Failed to create session: ${String(event.data.message ?? "")}`);
    }
    throw new Error(`Unexpected response: ${event.type}`);
  }

  async resumeSession(sessionId: number): Promise<boolean> {
    await this.connect();
    await this.send({ type: "resume_session", session_id: sessionId });
    const event = await this.receive();
    if (event.type === "session_ready") {
      const id = Number(event.data.session_id);
      this.currentSessionId = Number.isFinite(id) ? id : sessionId;
      return true;
    }
    if (event.type === "error") {
      console.warn(`Failed to resume session ${sessionId}: ${String(event.data.message ?? "")}`);
      return false;
    }
    throw new Error(`Unexpected response: ${event.type}`);
  }

  async updateConfig(config: SessionConfig): Promise<boolean> {
    if (!this.currentSessionId) {
      throw new Error("No active session to update");
    }
    await this.send({ type: "update_config", config });
    const event = await this.receive();
    if (event.type === "session_ready") {
      return true;
    }
    if (event.type === "error") {
      console.warn(`Failed to update config: ${String(event.data.message ?? "")}`);
      return false;
    }
    throw new Error(`Unexpected response: ${event.type}`);
  }

  async *query(prompt: string): AsyncIterable<StreamEvent> {
    if (!this.currentSessionId) {
      throw new Error("No active session. Call newSession or resumeSession first.");
    }
    await this.send({ type: "query", prompt });
    while (true) {
      const event = await this.receive();
      yield event;
      if (event.type === "error" && event.data.recoverable === false) {
        break;
      }
      if (event.type === "done") {
        break;
      }
    }
  }

  async ensureSession(config: SessionConfig, sessionId?: number | null): Promise<number> {
    await this.connect();
    if (sessionId) {
      const resumed = await this.resumeSession(sessionId);
      if (resumed && this.currentSessionId) {
        return this.currentSessionId;
      }
    }
    return this.newSession(config);
  }
}
